File: Pharmaceutical_Company_Example/Structural/Python/Bridge.py
# Abstraction
class RegulatoryProcess:
    def __init__(self, implementation):
        self.implementation = implementation

    def submit(self):
        self.implementation.submit()

    def approve(self):
        self.implementation.approve()


# Concrete implementation for Vaccines
class VaccineProcess:
    def submit(self):
        print("Submitting vaccine clinical trial data...")

    def approve(self):
        print("Vaccine regulatory approval granted.")


# Concrete implementation for Small-Molecule Drugs
class SmallMoleculeDrugProcess:
    def submit(self):
        print("Submitting small-molecule clinical trial data...")

    def approve(self):
        print("Small-molecule regulatory approval granted.")


# Concrete implementation for Biologics
class BiologicsProcess:
    def submit(self):
        print("Submitting biologics clinical trial data...")

    def approve(self):
        print("Biologics regulatory approval granted.")


preclinical_small_molecule = RegulatoryProcess(SmallMoleculeDrugProcess())
preclinical_small_molecule.submit()
preclinical_small_molecule.approve()

clinical_biologics = RegulatoryProcess(SmallMoleculeDrugProcess())
clinical_biologics.submit()
clinical_biologics.approve()

clinical_vaccine = RegulatoryProcess(VaccineProcess())
clinical_vaccine.submit()
clinical_vaccine.approve()

File: Pharmaceutical_Company_Example/Structural/R/Bridge.R

# Abstraction
RegulatoryProcess <- setRefClass(
  "RegulatoryProcess",
  fields = list(implementation = "ANY"),
  methods = list(
    initialize = function(implementation) {
      .self$implementation <- implementation
    },
    submit = function() {
      .self$implementation$submit()
    },
    approve = function() {
      .self$implementation$approve()
    }
  )
)

# Concrete implementation for Vaccines
VaccineProcess <- setRefClass(
  "VaccineProcess",
  methods = list(
    submit = function() {
      message("Submitting vaccine clinical trial data...")
    },
    approve = function() {
      message("Vaccine regulatory approval granted.")
    }
  )
)

# Concrete implementation for Small-Molecule Drugs
SmallMoleculeDrugProcess <- setRefClass(
  "SmallMoleculeDrugProcess",
  methods = list(
    submit = function() {
      message("Submitting small-molecule clinical trial data...")
    },
    approve = function() {
      message("Small-molecule regulatory approval granted.")
    }
  )
)

# Concrete implementation for Biologics
BiologicsProcess <- setRefClass(
  "BiologicsProcess",
  methods = list(
    submit = function() {
      message("Submitting biologics clinical trial data...")
    },
    approve = function() {
      message("Biologics regulatory approval granted.")
    }
  )
)


preclinical_small_molecule <- RegulatoryProcess$new(SmallMoleculeDrugProcess$new())
preclinical_small_molecule$submit()
preclinical_small_molecule$approve()

clinical_biologics <- RegulatoryProcess$new(SmallMoleculeDrugProcess$new())
clinical_biologics$submit()
clinical_biologics$approve()

clinical_Vaccine <- RegulatoryProcess$new(VaccineProcess$new())
clinical_Vaccine$submit()
clinical_Vaccine$approve()

